components/kyma-environment-broker/internal/runtime/handler.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage/dberr"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+)
+
+const (
+	limitParam  = "limit"
+	cursorParam = "cursor"
+)
+
+//go:generate mockery -name=Converter -output=automock -outpkg=automock -case=underscore
+type Converter interface {
+	InstancesAndOperationsToDTO(internal.Instance, *internal.ProvisioningOperation, *internal.DeprovisioningOperation, *internal.UpgradeKymaOperation) (RuntimeDTO, error)
+}
+
+type Handler struct {
+	instancesDb  storage.Instances
+	operationsDb storage.Operations
+	converter    Converter
+
+	defaultMaxPage int
+}
+
+func NewHandler(instanceDb storage.Instances, operationDb storage.Operations, defaultMaxPage int, converter Converter) *Handler {
+	return &Handler{
+		instancesDb:    instanceDb,
+		operationsDb:   operationDb,
+		converter:      converter,
+		defaultMaxPage: defaultMaxPage,
+	}
+}
+
+func (h *Handler) AttachRoutes(router *mux.Router) {
+	router.HandleFunc("/runtimes", h.getRuntimes)
+}
+
+func (h *Handler) getRuntimes(w http.ResponseWriter, req *http.Request) {
+	var toReturn []RuntimeDTO
+	limit, cursor, err := h.getParams(req)
+	if err != nil {
+		httputil.WriteErrorResponse(w, http.StatusBadRequest, errors.Wrap(err, "while getting query parameters"))
+		return
+	}
+
+	instances, pageInfo, totalCount, err := h.instancesDb.List(limit, cursor)
+	if err != nil {
+		httputil.WriteErrorResponse(w, http.StatusInternalServerError, errors.Wrap(err, "while fetching instances"))
+		return
+	}
+
+	for _, instance := range instances {
+		pOpr, dOpr, ukOpr, err := h.getOperationsForInstance(instance)
+		if err != nil {
+			httputil.WriteErrorResponse(w, http.StatusInternalServerError, errors.Wrap(err, "while fetching operations for instance"))
+			return
+		}
+
+		dto, err := h.converter.InstancesAndOperationsToDTO(instance, pOpr, dOpr, ukOpr)
+		if err != nil {
+			httputil.WriteErrorResponse(w, http.StatusInternalServerError, errors.Wrap(err, "while converting instances to DTO"))
+			return
+		}
+
+		toReturn = append(toReturn, dto)
+	}
+
+	page := RuntimesPage{
+		Data:       toReturn,
+		PageInfo:   pageInfo,
+		TotalCount: totalCount,
+	}
+	httputil.WriteResponse(w, http.StatusOK, page)
+}
+
+func (h *Handler) getOperationsForInstance(instance internal.Instance) (*internal.ProvisioningOperation, *internal.DeprovisioningOperation, *internal.UpgradeKymaOperation, error) {
+	pOpr, err := h.operationsDb.GetProvisioningOperationByInstanceID(instance.InstanceID)
+	if err != nil && !dberr.IsNotFound(err) {
+		return nil, nil, nil, err
+	}
+	dOpr, err := h.operationsDb.GetDeprovisioningOperationByInstanceID(instance.InstanceID)
+	if err != nil && !dberr.IsNotFound(err) {
+		return nil, nil, nil, err
+	}
+	ukOpr, err := h.operationsDb.GetUpgradeKymaOperationByInstanceID(instance.InstanceID)
```

Comment:
We can assume that there is at most one provisioning and one deprovisioning operation per instance. But it is nothing strange to see more than one upgrade operation per instance. Please review the storage operation